sdk/ml/azure-ai-ml/azure/ai/ml/_utils/_cache_utils.py

```diff
@@ -0,0 +1,76 @@
+# ---------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# ---------------------------------------------------------
+import uuid
+from collections import defaultdict
+from typing import Union, Tuple, List
+
+from azure.ai.ml.constants._common import AzureMLResourceType
+from azure.ai.ml.entities import Component
+from azure.ai.ml.entities._builders import BaseNode
+
+
+class CachedNodeResolver(object):
+    """Class to resolve component in nodes with cached component resolution results.
+    """
+    _ANONYMOUS_HASH_PREFIX = "anonymous_component:"
+    _YAML_SOURCE_PREFIX = "yaml_source:"
+    _ORIGIN_PREFIX = "origin:"
+
+    def __init__(self, resolver):
+        self._resolver = resolver
+        self._component_resolution_cache = {}
+        self._new_components: List[Tuple[str, Component]] = []
+        self._nodes_to_apply = defaultdict(list)
+
+    def _resolve_component(self, component: Union[str, Component]) -> str:
+        return self._resolver(component, azureml_type=AzureMLResourceType.COMPONENT)
+
+    @classmethod
+    def _get_component_hash(cls, component: Union[Component, str]):
+        """Get a hash for a component."""
+        if isinstance(component, str):
+            # component can be arm string like "train_model:1"
+            return cls._ORIGIN_PREFIX + component
+
+        # TODO: add hash function here
+        # for pipeline component, all dependencies of pipeline component has already been resolved
+        # so we can safely use its anonymous hash
+        # for non-pipeline component, we can't directly use anonymous hash as code hasn't been uploaded yet
+        # return a random hash by default
+        return cls._ANONYMOUS_HASH_PREFIX + str(uuid.uuid4())
+
+    def register_node_to_resolve(self, node: BaseNode):
```

Comment:
Is this one thread safe?

Reply:
Operations on attributes of `CachedNodeResolver` should be thread safe as they're not shared among threads.

Risk is on hash calculation part. The key is to make sure that the input `node` won't change on & after this function call. Here we not only resolve components layer by layer, but also register nodes layer by layer, which means that all dependencies of the input `node` has already been resolved.

Reply:
on the other hand, if 2 nodes in the same layer share the same component, node update (`node._component = resolved_component`) will be put off till all registrations done, so still thread safe.

Reply:
Yes. In the way we use this class, there is no thread-safe problem. 
- resolve_nodes & register_node_to_resolve will not be called concurrently
- nodes registered will be independent

May document the usage assumption.

Reply:
Updated docstring for the class